// services/update-topic/update-topic-response.js
"use strict";
/**
 * @module Services.UpdateTopic
 */
Object.defineProperty(exports, "__esModule", { value: true });
var enumerize_1 = require("util/enumerize");
var topics_1 = require("../../../topics/topics");
// tslint:disable-next-line:variable-name
exports.UpdateTopicResponseEnum = {
    SUCCESS: {
        id: 0,
        reason: null
    },
    INCOMPATIBLE_UPDATE: {
        id: 1,
        reason: topics_1.UpdateFailReasonEnum.INCOMPATIBLE_UPDATE
    },
    UPDATE_FAILED: {
        id: 2,
        reason: topics_1.UpdateFailReasonEnum.UPDATE_FAILED
    },
    INVALID_UPDATER: {
        id: 3,
        reason: topics_1.UpdateFailReasonEnum.INVALID_UPDATER
    },
    MISSING_TOPIC: {
        id: 4,
        reason: topics_1.UpdateFailReasonEnum.MISSING_TOPIC
    },
    EXCLUSIVE_UPDATER_CONFLICT: {
        id: 7,
        reason: topics_1.UpdateFailReasonEnum.EXCLUSIVE_UPDATER_CONFLICT
    },
    DELTA_WITHOUT_VALUE: {
        id: 8,
        reason: topics_1.UpdateFailReasonEnum.DELTA_WITHOUT_VALUE
    },
    CLUSTER_REPARTITION: {
        id: 9,
        reason: topics_1.UpdateFailReasonEnum.CLUSTER_REPARTITION
    },
    INCOMPATIBLE_STATE: {
        id: 10,
        reason: topics_1.UpdateFailReasonEnum.INCOMPATIBLE_STATE
    },
    UNSATISFIED_CONSTRAINT: {
        id: -1,
        reason: topics_1.UpdateFailReasonEnum.UNSATISFIED_CONSTRAINT
    },
    INVALID_UPDATE_STREAM: {
        id: -2,
        reason: topics_1.UpdateFailReasonEnum.INVALID_UPDATE_STREAM
    }
};
enumerize_1.enumerize(exports.UpdateTopicResponseEnum);
Object.freeze(exports.UpdateTopicResponseEnum);
/**
 * Server response from a UPDATE_TOPIC_SET request.
 */
var UpdateTopicResponse = /** @class */ (function () {
    /**
     * Create a new UpdateSourceDeltaRequest instance
     *
     * @param reason  the reason for failure if there was an error
     * @param isError  a flag indicating whether there was an error
     */
    function UpdateTopicResponse(reason, isError) {
        this.reason = reason;
        this.isError = isError;
    }
    return UpdateTopicResponse;
}());
exports.UpdateTopicResponse = UpdateTopicResponse;
