// services/client/client-location-serialiser.js
"use strict";
/**
 * @module Services.Client
 */
Object.defineProperty(exports, "__esModule", { value: true });
var Codec = require("io/codec");
var BEES = require("serialisers/byte-encoded-enum-serialiser");
var math_1 = require("util/math");
var client_location_1 = require("../../../client/client-location");
/**
 * Read a {@link ClientLocation} from the stream
 *
 * @param bis  the input stream
 * @return     the {@link ClientLocation} that was read
 */
function read(bis) {
    var address = Codec.readString(bis);
    var hostname = Codec.readString(bis) || address;
    var resolved = Codec.readString(bis) || hostname;
    var addressType = BEES.read(bis, client_location_1.AddressType);
    var details = {
        country: Codec.readString(bis),
        language: Codec.readString(bis)
    };
    var coordinates = {
        latitude: math_1.intBitsToFloat(Codec.readInt32(bis)),
        longitude: math_1.intBitsToFloat(Codec.readInt32(bis))
    };
    return {
        type: addressType,
        address: address,
        hostname: hostname,
        resolved: resolved,
        details: details,
        coordinates: coordinates
    };
}
exports.read = read;
