// features/topics/value-stream-adapter.js
"use strict";
/**
 * @module ValueStream
 */
Object.defineProperty(exports, "__esModule", { value: true });
var datatypes_1 = require("data/datatypes");
/**
 * Adapter for transforming the type of the emitted value to the datatype of
 * the ValueStream.
 */
var ValueStreamAdapter = /** @class */ (function () {
    /**
     * Create a new ValueStreamAdapter instance
     *
     * @param stream        the stream that is wrapped
     * @param datatype      the data type of the value stream
     * @param topictype     the topic type
     * @param subscription  the subscription receiving the data
     * @param emitter       the emitter of the underlying stream adapter
     */
    function ValueStreamAdapter(stream, datatype, topictype) {
        this.stream = stream;
        this.datatype = datatype;
        this.sourceDatatype = datatypes_1.DataTypes.getByValue(topictype);
    }
    /**
     * @inheritdoc
     */
    ValueStreamAdapter.prototype.selects = function (specification) {
        return this.stream.selects(specification);
    };
    /**
     * @inheritdoc
     */
    ValueStreamAdapter.prototype.onOpen = function () {
        return this.stream.onOpen();
    };
    /**
     * @inheritdoc
     */
    ValueStreamAdapter.prototype.onDelta = function (path, specification, oldContent, content, delta, oldValue, newValue) {
        return this.stream.onDelta(path, specification, oldContent, content, delta, (oldContent !== null)
            ? this.sourceDatatype.readAs(this.datatype.valueClass, oldContent)
            : null, this.sourceDatatype.readAs(this.datatype.valueClass, content));
    };
    /**
     * @inheritdoc
     */
    ValueStreamAdapter.prototype.onValue = function (path, specification, oldContent, content, oldValue, newValue) {
        return this.stream.onValue(path, specification, oldContent, content, (oldContent !== null)
            ? this.sourceDatatype.readAs(this.datatype.valueClass, oldContent)
            : null, this.sourceDatatype.readAs(this.datatype.valueClass, content));
    };
    /**
     * @inheritdoc
     */
    ValueStreamAdapter.prototype.onSubscription = function (path, specification) {
        return this.stream.onSubscription(path, specification);
    };
    /**
     * @inheritdoc
     */
    ValueStreamAdapter.prototype.onUnsubscription = function (path, specification, reason) {
        return this.stream.onUnsubscription(path, specification, reason);
    };
    /**
     * @inheritdoc
     */
    ValueStreamAdapter.prototype.onSubscriptionError = function (error) {
        return this.stream.onSubscriptionError(error);
    };
    return ValueStreamAdapter;
}());
exports.ValueStreamAdapter = ValueStreamAdapter;
