// services/update-source/update-source-update-response.js
"use strict";
/**
 * @module Services.UpdateSource
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Server response code from a UPDATE_SOURCE_SET or UPDATE_SOURCE_DELTA
 */
var ResponseCode;
(function (ResponseCode) {
    ResponseCode[ResponseCode["SUCCESS"] = 0] = "SUCCESS";
    ResponseCode[ResponseCode["INCOMPATIBLE_UPDATE"] = 1] = "INCOMPATIBLE_UPDATE";
    ResponseCode[ResponseCode["UPDATE_FAILED"] = 2] = "UPDATE_FAILED";
    ResponseCode[ResponseCode["INVALID_UPDATER"] = 3] = "INVALID_UPDATER";
    ResponseCode[ResponseCode["MISSING_TOPIC"] = 4] = "MISSING_TOPIC";
    ResponseCode[ResponseCode["INVALID_ADDRESS"] = 5] = "INVALID_ADDRESS";
    ResponseCode[ResponseCode["DUPLICATES"] = 6] = "DUPLICATES";
    ResponseCode[ResponseCode["EXCLUSIVE_UPDATER_CONFLICT"] = 7] = "EXCLUSIVE_UPDATER_CONFLICT";
    ResponseCode[ResponseCode["DELTA_WITHOUT_VALUE"] = 8] = "DELTA_WITHOUT_VALUE";
    ResponseCode[ResponseCode["CLUSTER_REPARTITION"] = 9] = "CLUSTER_REPARTITION";
    ResponseCode[ResponseCode["INCOMPATIBLE_STATE"] = 10] = "INCOMPATIBLE_STATE";
})(ResponseCode = exports.ResponseCode || (exports.ResponseCode = {}));
/**
 * Server response from a UPDATE_SOURCE_SET or UPDATE_SOURCE_DELTA request
 */
var UpdateSourceUpdateResponse = /** @class */ (function () {
    /**
     * Create a new UpdateSourceUpdateResponse instance
     *
     * @param error  the error code if there was an error
     */
    function UpdateSourceUpdateResponse(error) {
        this.error = error;
    }
    return UpdateSourceUpdateResponse;
}());
exports.UpdateSourceUpdateResponse = UpdateSourceUpdateResponse;
