// services/services.js
"use strict";
/**
 * @module Services
 */
Object.defineProperty(exports, "__esModule", { value: true });
var topic_specification_info_serialiser_1 = require("topics/details/topic-specification-info-serialiser");
var topic_selector_serialiser_1 = require("topics/topic-selector-serialiser");
var boolean_serialiser_1 = require("serialisers/boolean-serialiser");
var bytes_serialiser_1 = require("serialisers/bytes-serialiser");
var string_serialiser_1 = require("serialisers/string-serialiser");
var security_command_script_serialiser_1 = require("services/authentication/security-command-script-serialiser");
var system_authentication_configuration_serialiser_1 = require("services/authentication/system-authentication-configuration-serialiser");
var change_principal_request_serialiser_1 = require("services/change-principal/change-principal-request-serialiser");
var client_filter_subscribe_request_serialiser_1 = require("services/control-client/client-filter-subscribe-request-serialiser");
var client_filter_subscribe_response_serialiser_1 = require("services/control-client/client-filter-subscribe-response-serialiser");
var client_subscribe_request_serialiser_1 = require("services/control-client/client-subscribe-request-serialiser");
var change_authorisation_roles_filter_request_serialiser_1 = require("services/control/change-authorisation-roles-filter-request-serialiser");
var change_authorisation_roles_request_serialiser_1 = require("services/control/change-authorisation-roles-request-serialiser");
var close_client_request_serialiser_1 = require("services/control/close-client-request-serialiser");
var get_session_properties_request_serialiser_1 = require("services/control/get-session-properties-request-serialiser");
var get_session_properties_response_serialiser_1 = require("services/control/get-session-properties-response-serialiser");
var message_receiver_control_registration_params_serialiser_1 = require("services/control/message-receiver-control-registration-params-serialiser");
var message_receiver_control_registration_request_serialiser_1 = require("services/control/message-receiver-control-registration-request-serialiser");
var messaging_client_forward_send_request_serialiser_1 = require("services/control/messaging-client-forward-send-request-serialiser");
var messaging_client_send_request_serialiser_1 = require("services/control/messaging-client-send-request-serialiser");
var messaging_filter_response_serialiser_1 = require("services/control/messaging-filter-response-serialiser");
var messaging_filter_send_request_serialiser_1 = require("services/control/messaging-filter-send-request-serialiser");
var messaging_filter_send_response_serialiser_1 = require("services/control/messaging-filter-send-response-serialiser");
var messaging_response_serialiser_1 = require("services/control/messaging-response-serialiser");
var messaging_send_request_serialiser_1 = require("services/control/messaging-send-request-serialiser");
var session_properties_event_batch_serialiser_1 = require("services/control/session-properties-event-batch-serialiser");
var set_client_conflation_filter_request_serialiser_1 = require("services/control/set-client-conflation-filter-request-serialiser");
var set_client_conflation_request_serialiser_1 = require("services/control/set-client-conflation-request-serialiser");
var set_session_properties_filter_request_serialiser_1 = require("services/control/set-session-properties-filter-request-serialiser");
var set_session_properties_listener_serialiser_1 = require("services/control/set-session-properties-listener-serialiser");
var set_session_properties_request_serialiser_1 = require("services/control/set-session-properties-request-serialiser");
var set_session_properties_result_serialiser_1 = require("services/control/set-session-properties-result-serialiser");
var topic_control_registration_params_serialiser_1 = require("services/control/topic-control-registration-params-serialiser");
var topic_control_registration_request_serialiser_1 = require("services/control/topic-control-registration-request-serialiser");
var count_or_parser_errors_serialiser_1 = require("services/count-or-parser-errors-serialiser");
var error_report_list_serialiser_1 = require("services/error-report-list-serialiser");
var jmx_object_names_serialiser_1 = require("services/fetch-jmx-values/jmx-object-names-serialiser");
var mbean_values_serialiser_1 = require("services/fetch-jmx-values/mbean-values-serialiser");
var fetch_query_result_serialiser_1 = require("services/fetch-query/fetch-query-result-serialiser");
var fetch_query_serialiser_1 = require("services/fetch-query/fetch-query-serialiser");
var fetch_request_serialiser_1 = require("services/fetch/fetch-request-serialiser");
var entries_fetch_request_serialiser_1 = require("services/log-retrieval/entries-fetch-request-serialiser");
var entries_fetch_response_serialiser_1 = require("services/log-retrieval/entries-fetch-response-serialiser");
var session_metric_collector_list_serialiser_1 = require("services/metric-collectors/session-metric-collector-list-serialiser");
var session_metric_collector_serialiser_1 = require("services/metric-collectors/session-metric-collector-serialiser");
var topic_metric_collector_list_serialiser_1 = require("services/metric-collectors/topic-metric-collector-list-serialiser");
var topic_metric_collector_serialiser_1 = require("services/metric-collectors/topic-metric-collector-serialiser");
var measured_entity_class_metrics_serialiser_1 = require("services/metrics/measured-entity-class-metrics-serialiser");
var server_metrics_serialiser_1 = require("services/metrics/server-metrics-serialiser");
var missing_topic_request_serialiser_1 = require("services/missing-topic/missing-topic-request-serialiser");
var check_remote_server_result_serialiser_1 = require("services/remote-servers/check-remote-server-result-serialiser");
var create_remote_server_result_serialiser_1 = require("services/remote-servers/create-remote-server-result-serialiser");
var list_remote_servers_result_serialiser_1 = require("services/remote-servers/list-remote-servers-result-serialiser");
var remote_server_definition_serialiser_1 = require("services/remote-servers/remote-server-definition-serialiser");
var remove_topic_serialiser_1 = require("services/remove-topic/remove-topic-serialiser");
var replace_licence_response_serialiser_1 = require("services/replace-licence/replace-licence-response-serialiser");
var authentication_control_registration_parameters_serialiser_1 = require("services/security/authentication-control-registration-parameters-serialiser");
var authentication_control_registration_request_serialiser_1 = require("services/security/authentication-control-registration-request-serialiser");
var authentication_request_serialiser_1 = require("services/security/authentication-request-serialiser");
var authentication_response_serialiser_1 = require("services/security/authentication-response-serialiser");
var authenticator_registration_parameters_serialiser_1 = require("services/security/authenticator-registration-parameters-serialiser");
var authenticator_registration_request_serialiser_1 = require("services/security/authenticator-registration-request-serialiser");
var authenticator_request_serialiser_1 = require("services/security/authenticator-request-serialiser");
var authenticator_response_serialiser_1 = require("services/security/authenticator-response-serialiser");
var list_global_permissions_result_serialiser_1 = require("services/security/list-global-permissions-result-serialiser");
var list_path_permissions_result_serialiser_1 = require("services/security/list-path-permissions-result-serialiser");
var security_configuration_serialiser_1 = require("services/security/security-configuration-serialiser");
var session_lock_acquisition_serialiser_1 = require("services/session-lock/session-lock-acquisition-serialiser");
var session_lock_request_serialiser_1 = require("services/session-lock/session-lock-request-serialiser");
var range_query_request_serialiser_1 = require("services/timeseries/range-query-request-serialiser");
var range_query_result_serialiser_1 = require("services/timeseries/range-query-result-serialiser");
var time_series_append_request_serialiser_1 = require("services/timeseries/time-series-append-request-serialiser");
var time_series_edit_request_serialiser_1 = require("services/timeseries/time-series-edit-request-serialiser");
var time_series_event_metadata_serialiser_1 = require("services/timeseries/time-series-event-metadata-serialiser");
var topic_add_request_serialiser_1 = require("services/topic-add/topic-add-request-serialiser");
var topic_add_response_serialiser_1 = require("services/topic-add/topic-add-response-serialiser");
var topic_notification_deregistration_request_serialiser_1 = require("services/topic-notifications/topic-notification-deregistration-request-serialiser");
var topic_notification_descendant_event_serialiser_1 = require("services/topic-notifications/topic-notification-descendant-event-serialiser");
var topic_notification_deselection_serialiser_1 = require("services/topic-notifications/topic-notification-deselection-serialiser");
var topic_notification_event_serialiser_1 = require("services/topic-notifications/topic-notification-event-serialiser");
var topic_notification_selection_serialiser_1 = require("services/topic-notifications/topic-notification-selection-serialiser");
var add_and_set_topic_request_serialiser_1 = require("services/topic-update/add-and-set-topic-request-serialiser");
var add_topic_result_serialiser_1 = require("services/topic-update/add-topic-result-serialiser");
var apply_json_patch_request_serialiser_1 = require("services/topic-update/apply-json-patch-request-serialiser");
var create_update_stream_and_set_request_serialiser_1 = require("services/topic-update/create-update-stream-and-set-request-serialiser");
var create_update_stream_request_serialiser_1 = require("services/topic-update/create-update-stream-request-serialiser");
var json_patch_result_serialiser_1 = require("services/topic-update/json-patch-result-serialiser");
var set_topic_request_serialiser_1 = require("services/topic-update/set-topic-request-serialiser");
var update_stream_add_topic_request_serialiser_1 = require("services/topic-update/update-stream-add-topic-request-serialiser");
var update_stream_add_topic_response_serialiser_1 = require("services/topic-update/update-stream-add-topic-response-serialiser");
var update_stream_id_serialiser_1 = require("services/topic-update/update-stream-id-serialiser");
var update_stream_request_serialiser_1 = require("services/topic-update/update-stream-request-serialiser");
var create_topic_view_result_serialiser_1 = require("services/topic-views/create-topic-view-result-serialiser");
var list_topic_views_result_serialiser_1 = require("services/topic-views/list-topic-views-result-serialiser");
var named_topic_view_specification_serialiser_1 = require("services/topic-views/named-topic-view-specification-serialiser");
var unsubscription_notification_serialiser_1 = require("services/unsubscription-notification/unsubscription-notification-serialiser");
var update_source_delta_request_serialiser_1 = require("services/update-source/update-source-delta-request-serialiser");
var update_source_deregistration_request_serialiser_1 = require("services/update-source/update-source-deregistration-request-serialiser");
var update_source_registration_request_serialiser_1 = require("services/update-source/update-source-registration-request-serialiser");
var update_source_set_request_serialiser_1 = require("services/update-source/update-source-set-request-serialiser");
var update_source_state_request_serialiser_1 = require("services/update-source/update-source-state-request-serialiser");
var update_source_state_serialiser_1 = require("services/update-source/update-source-state-serialiser");
var update_source_update_response_serialiser_1 = require("services/update-source/update-source-update-response-serialiser");
var update_topic_response_serialiser_1 = require("services/update-topic/update-topic-response-serialiser");
var update_topic_set_request_serialiser_1 = require("services/update-topic/update-topic-set-request-serialiser");
var topic_will_parameters_serialiser_1 = require("services/wills/topic-will-parameters-serialiser");
var will_registration_result_serialiser_1 = require("services/wills/will-registration-result-serialiser");
// tslint:disable-next-line:variable-name
var NullSerialiser = {
    read: function () { return null; },
    write: function () { }
};
exports.FETCH = {
    id: 2,
    request: fetch_request_serialiser_1.FetchRequestSerialiser,
    response: NullSerialiser
};
exports.SUBSCRIBE = {
    id: 3,
    request: topic_selector_serialiser_1.TopicSelectorSerialiser,
    response: NullSerialiser
};
exports.UNSUBSCRIBE = {
    id: 4,
    request: topic_selector_serialiser_1.TopicSelectorSerialiser,
    response: NullSerialiser
};
exports.SUBSCRIBE_CLIENT = {
    id: 10,
    request: client_subscribe_request_serialiser_1.ClientSubscribeRequestSerialiser,
    response: NullSerialiser
};
exports.UNSUBSCRIBE_CLIENT = {
    id: 11,
    request: client_subscribe_request_serialiser_1.ClientSubscribeRequestSerialiser,
    response: NullSerialiser
};
exports.CLOSE_CLIENT = {
    id: 14,
    request: close_client_request_serialiser_1.CloseClientRequestSerialiser,
    response: NullSerialiser
};
exports.SET_CLIENT_CONFLATION = {
    id: 15,
    request: set_client_conflation_request_serialiser_1.SetClientConflationRequestSerialiser,
    response: NullSerialiser
};
exports.UNSUBSCRIPTION_NOTIFICATION = {
    id: 42,
    request: unsubscription_notification_serialiser_1.UnsubscriptionNotificationSerialiser,
    response: NullSerialiser
};
exports.MISSING_TOPIC = {
    id: 50,
    request: missing_topic_request_serialiser_1.MissingTopicRequestSerialiser,
    response: boolean_serialiser_1.BooleanSerialiser
};
exports.TOPIC_SCOPED_WILL_REGISTRATION = {
    id: 53,
    request: topic_will_parameters_serialiser_1.TopicWillParametersSerialiser,
    response: will_registration_result_serialiser_1.WillRegistrationResultSerialiser
};
exports.TOPIC_SCOPED_WILL_DEREGISTRATION = {
    id: 54,
    request: topic_will_parameters_serialiser_1.TopicWillParametersSerialiser,
    response: NullSerialiser
};
exports.SYSTEM_PING = {
    id: 55,
    request: NullSerialiser,
    response: NullSerialiser
};
exports.USER_PING = {
    id: 56,
    request: NullSerialiser,
    response: NullSerialiser
};
exports.CHANGE_PRINCIPAL = {
    id: 5,
    request: change_principal_request_serialiser_1.ChangePrincipalRequestSerialiser,
    response: boolean_serialiser_1.BooleanSerialiser
};
exports.GET_SYSTEM_AUTHENTICATION = {
    id: 57,
    request: NullSerialiser,
    response: system_authentication_configuration_serialiser_1.SystemAuthenticationConfigurationSerialiser
};
exports.UPDATE_SYSTEM_AUTHENTICATION = {
    id: 58,
    request: security_command_script_serialiser_1.SecurityCommandScriptSerialiser,
    response: error_report_list_serialiser_1.ErrorReportListSerialiser
};
exports.GET_SECURITY_CONFIGURATION = {
    id: 59,
    request: NullSerialiser,
    response: security_configuration_serialiser_1.SecurityConfigurationSerialiser
};
exports.UPDATE_SECURITY_CONFIGURATION = {
    id: 60,
    request: security_command_script_serialiser_1.SecurityCommandScriptSerialiser,
    response: error_report_list_serialiser_1.ErrorReportListSerialiser
};
exports.TOPIC_CONTROL_REGISTRATION = {
    id: 20,
    request: topic_control_registration_request_serialiser_1.TopicControlRegistrationRequestSerialiser,
    response: NullSerialiser
};
exports.TOPIC_CONTROL_DEREGISTRATION = {
    id: 21,
    request: topic_control_registration_params_serialiser_1.TopicControlRegistrationParamsSerialiser,
    response: NullSerialiser
};
exports.AUTHENTICATION_CONTROL_REGISTRATION = {
    id: 22,
    request: authentication_control_registration_request_serialiser_1.AuthenticationControlRegistrationRequestSerialiser,
    response: NullSerialiser
};
exports.AUTHENTICATION_CONTROL_DEREGISTRATION = {
    id: 23,
    request: authentication_control_registration_parameters_serialiser_1.AuthenticationControlRegistrationParametersSerialiser,
    response: NullSerialiser
};
exports.AUTHENTICATION = {
    id: 24,
    request: authentication_request_serialiser_1.AuthenticationRequestSerialiser,
    response: authentication_response_serialiser_1.AuthenticationResponseSerialiser
};
exports.AUTHENTICATOR_REGISTRATION = {
    id: 119,
    request: authenticator_registration_request_serialiser_1.AuthenticatorRegistrationRequestSerialiser,
    response: NullSerialiser
};
exports.AUTHENTICATOR = {
    id: 120,
    request: authenticator_request_serialiser_1.AuthenticatorRequestSerialiser,
    response: authenticator_response_serialiser_1.AuthenticatorResponseSerialiser
};
exports.AUTHENTICATOR_DEREGISTRATION = {
    id: 121,
    request: authenticator_registration_parameters_serialiser_1.AuthenticatorRegistrationParametersSerialiser,
    response: NullSerialiser
};
exports.UPDATE_SOURCE_REGISTRATION = {
    id: 30,
    request: update_source_registration_request_serialiser_1.UpdateSourceRegistrationRequestSerialiser,
    response: update_source_state_serialiser_1.UpdateSourceStateSerialiser
};
exports.UPDATE_SOURCE_DEREGISTRATION = {
    id: 31,
    request: update_source_deregistration_request_serialiser_1.UpdateSourceDeregistrationRequestSerialiser,
    response: NullSerialiser
};
exports.UPDATE_SOURCE_STATE = {
    id: 32,
    request: update_source_state_request_serialiser_1.UpdateSourceStateRequestSerialiser,
    response: NullSerialiser
};
exports.MESSAGE_RECEIVER_CONTROL_REGISTRATION = {
    id: 63,
    request: message_receiver_control_registration_request_serialiser_1.MessageReceiverControlRegistrationRequestSerialiser,
    response: NullSerialiser
};
exports.MESSAGE_RECEIVER_CONTROL_DEREGISTRATION = {
    id: 64,
    request: message_receiver_control_registration_params_serialiser_1.MessageReceiverControlRegistrationParamsSerialiser,
    response: NullSerialiser
};
exports.FILTER_SUBSCRIBE = {
    id: 65,
    request: client_filter_subscribe_request_serialiser_1.ClientFilterSubscribeRequestSerialiser,
    response: client_filter_subscribe_response_serialiser_1.ClientFilterSubscribeResponseSerialiser
};
exports.FILTER_UNSUBSCRIBE = {
    id: 66,
    request: client_filter_subscribe_request_serialiser_1.ClientFilterSubscribeRequestSerialiser,
    response: client_filter_subscribe_response_serialiser_1.ClientFilterSubscribeResponseSerialiser
};
exports.GET_SESSION_PROPERTIES = {
    id: 67,
    request: get_session_properties_request_serialiser_1.GetSessionPropertiesRequestSerialiser,
    response: get_session_properties_response_serialiser_1.GetSessionPropertiesResponseSerialiser
};
exports.SET_SESSION_PROPERTIES = {
    id: 105,
    request: set_session_properties_request_serialiser_1.SetSessionPropertiesRequestSerialiser,
    response: set_session_properties_result_serialiser_1.SetSessionPropertiesResultSerialiser
};
exports.SET_SESSION_PROPERTIES_FILTER = {
    id: 106,
    request: set_session_properties_filter_request_serialiser_1.SetSessionPropertiesFilterRequestSerialiser,
    response: count_or_parser_errors_serialiser_1.CountOrParserErrorsSerialiser
};
exports.MESSAGING_SEND = {
    id: 85,
    request: messaging_send_request_serialiser_1.MessagingSendRequestSerialiser,
    response: messaging_response_serialiser_1.MessagingResponseSerialiser
};
exports.MESSAGING_RECEIVER_SERVER = {
    id: 86,
    request: messaging_client_send_request_serialiser_1.MessagingClientSendRequestSerialiser,
    response: messaging_response_serialiser_1.MessagingResponseSerialiser
};
exports.NOTIFY_SUBSCRIPTION = {
    id: 87,
    request: topic_specification_info_serialiser_1.TopicSpecificationInfoSerialiser,
    response: NullSerialiser
};
exports.MESSAGING_RECEIVER_CLIENT = {
    id: 88,
    request: messaging_client_forward_send_request_serialiser_1.MessagingClientForwardSendRequestSerialiser,
    response: messaging_response_serialiser_1.MessagingResponseSerialiser
};
exports.MESSAGING_FILTER_SEND = {
    id: 102,
    request: messaging_filter_send_request_serialiser_1.MessagingFilterSendRequestSerialiser,
    response: messaging_filter_send_response_serialiser_1.MessagingFilterSendResponseSerialiser
};
exports.FILTER_RESPONSE = {
    id: 103,
    request: messaging_filter_response_serialiser_1.MessagingFilterResponseSerialiser,
    response: NullSerialiser
};
exports.TOPIC_NOTIFICATION_DEREGISTRATION = {
    id: 91,
    request: topic_notification_deregistration_request_serialiser_1.TopicNotificationDeregistrationRequestSerialiser,
    response: NullSerialiser
};
exports.TOPIC_NOTIFICATION_EVENTS = {
    id: 92,
    request: topic_notification_event_serialiser_1.TopicNotificationEventSerialiser,
    response: NullSerialiser
};
exports.TOPIC_DESCENDANT_EVENTS = {
    id: 93,
    request: topic_notification_descendant_event_serialiser_1.TopicNotificationDescendantEventSerialiser,
    response: NullSerialiser
};
exports.TOPIC_NOTIFICATION_SELECTION = {
    id: 94,
    request: topic_notification_selection_serialiser_1.TopicNotificationSelectionSerialiser,
    response: NullSerialiser
};
exports.TOPIC_NOTIFICATION_DESELECTION = {
    id: 95,
    request: topic_notification_deselection_serialiser_1.TopicNotificationDeselectionSerialiser,
    response: NullSerialiser
};
exports.MESSAGING_RECEIVER_CONTROL_REGISTRATION = {
    id: 97,
    request: message_receiver_control_registration_request_serialiser_1.MessageReceiverControlRegistrationRequestSerialiser,
    response: NullSerialiser
};
exports.MESSAGING_RECEIVER_CONTROL_DEREGISTRATION = {
    id: 98,
    request: message_receiver_control_registration_params_serialiser_1.MessageReceiverControlRegistrationParamsSerialiser,
    response: NullSerialiser
};
exports.UPDATE_SOURCE_SET = {
    id: 77,
    request: update_source_set_request_serialiser_1.UpdateSourceSetRequestSerialiser,
    response: update_source_update_response_serialiser_1.UpdateSourceUpdateResponseSerialiser
};
exports.UPDATE_SOURCE_DELTA = {
    id: 78,
    request: update_source_delta_request_serialiser_1.UpdateSourceDeltaRequestSerialiser,
    response: update_source_update_response_serialiser_1.UpdateSourceUpdateResponseSerialiser
};
exports.UPDATE_TOPIC_SET = {
    id: 79,
    request: update_topic_set_request_serialiser_1.UpdateTopicSetRequestSerialiser,
    response: update_topic_response_serialiser_1.UpdateTopicResponseSerialiser
};
exports.SESSION_PROPERTIES_REGISTRATION_2 = {
    id: 81,
    request: set_session_properties_listener_serialiser_1.SetSessionPropertiesListenerSerialiser,
    response: NullSerialiser
};
exports.SESSION_PROPERTIES_EVENT_2 = {
    id: 82,
    request: session_properties_event_batch_serialiser_1.SessionPropertiesEventBatchSerialiser,
    response: NullSerialiser
};
exports.TOPIC_REMOVAL = {
    id: 83,
    request: remove_topic_serialiser_1.RemoveTopicSerialiser,
    response: NullSerialiser
};
exports.RANGE_QUERY = {
    id: 84,
    request: range_query_request_serialiser_1.RangeQueryRequestSerialiser,
    response: range_query_result_serialiser_1.RangeQueryResultSerialiser
};
exports.TIME_SERIES_APPEND = {
    id: 99,
    request: time_series_append_request_serialiser_1.TimeSeriesAppendRequestSerialiser,
    response: time_series_event_metadata_serialiser_1.EventMetadataImplSerialiser
};
exports.TIME_SERIES_EDIT = {
    id: 100,
    request: time_series_edit_request_serialiser_1.TimeSeriesEditRequestSerialiser,
    response: time_series_event_metadata_serialiser_1.EventMetadataImplSerialiser
};
exports.TOPIC_ADD = {
    id: 112,
    request: topic_add_request_serialiser_1.TopicAddRequestSerialiser,
    response: topic_add_response_serialiser_1.TopicAddResponseSerialiser
};
exports.ACQUIRE_SESSION_LOCK = {
    id: 114,
    request: session_lock_request_serialiser_1.SessionLockRequestSerialiser,
    response: session_lock_acquisition_serialiser_1.SessionLockAcquisitionSerialiser
};
exports.RELEASE_SESSION_LOCK = {
    id: 116,
    request: session_lock_acquisition_serialiser_1.SessionLockAcquisitionSerialiser,
    response: boolean_serialiser_1.BooleanSerialiser
};
exports.FETCH_QUERY = {
    id: 117,
    request: fetch_query_serialiser_1.FetchQuerySerialiser,
    response: fetch_query_result_serialiser_1.FetchQueryResultSerialiser
};
exports.SET_TOPIC = {
    id: 118,
    request: set_topic_request_serialiser_1.SetTopicRequestSerialiser,
    response: NullSerialiser
};
exports.ADD_AND_SET_TOPIC = {
    id: 122,
    request: add_and_set_topic_request_serialiser_1.AddAndSetTopicRequestSerialiser,
    response: add_topic_result_serialiser_1.AddTopicResultSerialiser
};
exports.APPLY_JSON_PATCH = {
    id: 151,
    request: apply_json_patch_request_serialiser_1.ApplyJsonPatchRequestSerialiser,
    response: json_patch_result_serialiser_1.JsonPatchResultSerialiser
};
exports.CHANGE_AUTHORISATION_ROLES = {
    id: 123,
    request: change_authorisation_roles_request_serialiser_1.ChangeAuthorisationRolesRequestSerialiser,
    response: NullSerialiser
};
exports.CHANGE_AUTHORISATION_ROLES_FILTER = {
    id: 124,
    request: change_authorisation_roles_filter_request_serialiser_1.ChangeAuthorisationRolesFilterRequestSerialiser,
    response: count_or_parser_errors_serialiser_1.CountOrParserErrorsSerialiser
};
exports.CREATE_UPDATE_STREAM = {
    id: 125,
    request: create_update_stream_request_serialiser_1.CreateUpdateStreamRequestSerialiser,
    response: update_stream_id_serialiser_1.UpdateStreamIdSerialiser
};
exports.CREATE_UPDATE_STREAM_AND_SET = {
    id: 126,
    request: create_update_stream_and_set_request_serialiser_1.CreateUpdateStreamAndSetRequestSerialiser,
    response: update_stream_id_serialiser_1.UpdateStreamIdSerialiser
};
exports.CHECK_UPDATE_STREAM = {
    id: 127,
    request: update_stream_id_serialiser_1.UpdateStreamIdSerialiser,
    response: NullSerialiser
};
exports.STREAM_SET_TOPIC_SERVICE = {
    id: 128,
    request: update_stream_request_serialiser_1.UpdateStreamRequestSerialiser,
    response: NullSerialiser
};
exports.STREAM_APPLY_DELTA_SERVICE = {
    id: 129,
    request: update_stream_request_serialiser_1.UpdateStreamRequestSerialiser,
    response: NullSerialiser
};
exports.STREAM_ADD_TOPIC_SERVICE = {
    id: 130,
    request: update_stream_add_topic_request_serialiser_1.UpdateStreamAddTopicRequestSerialiser,
    response: update_stream_add_topic_response_serialiser_1.UpdateStreamAddTopicResponseSerialiser
};
exports.STREAM_ADD_AND_SET_TOPIC_SERVICE = {
    id: 131,
    request: add_and_set_topic_request_serialiser_1.AddAndSetTopicRequestSerialiser,
    response: update_stream_add_topic_response_serialiser_1.UpdateStreamAddTopicResponseSerialiser
};
exports.FETCH_LOG_ENTRIES_SERVICE = {
    id: 132,
    request: entries_fetch_request_serialiser_1.EntriesFetchRequestSerialiser,
    response: entries_fetch_response_serialiser_1.EntriesFetchResponseSerialiser
};
exports.PUT_SESSION_METRIC_COLLECTOR = {
    id: 135,
    request: session_metric_collector_serialiser_1.SessionMetricCollectorSerialiser,
    response: error_report_list_serialiser_1.ErrorReportListSerialiser
};
exports.REMOVE_SESSION_METRIC_COLLECTOR = {
    id: 136,
    request: string_serialiser_1.StringSerialiser,
    response: NullSerialiser
};
exports.FETCH_SESSION_METRIC_COLLECTORS = {
    id: 137,
    request: NullSerialiser,
    response: session_metric_collector_list_serialiser_1.SessionMetricCollectorListSerialiser
};
exports.CREATE_TOPIC_VIEW = {
    id: 138,
    request: named_topic_view_specification_serialiser_1.NamedTopicViewSpecificationSerialiser,
    response: create_topic_view_result_serialiser_1.CreateTopicViewResultSerialiser
};
exports.REMOVE_TOPIC_VIEW = {
    id: 139,
    request: string_serialiser_1.StringSerialiser,
    response: NullSerialiser
};
exports.LIST_TOPIC_VIEWS = {
    id: 141,
    request: NullSerialiser,
    response: list_topic_views_result_serialiser_1.ListTopicViewsResultSerialiser
};
exports.GET_SERVER_METRICS = {
    id: 142,
    request: NullSerialiser,
    response: server_metrics_serialiser_1.ServerMetricsSerialiser
};
exports.LIST_GLOBAL_PERMISSIONS = {
    id: 143,
    request: NullSerialiser,
    response: list_global_permissions_result_serialiser_1.ListGlobalPermissionsResultSerialiser
};
exports.LIST_PATH_PERMISSIONS = {
    id: 144,
    request: string_serialiser_1.StringSerialiser,
    response: list_path_permissions_result_serialiser_1.ListPathPermissionsResultSerialiser
};
exports.GET_MEASURED_ENTITY_CLASS_METRICS = {
    id: 145,
    request: string_serialiser_1.StringSerialiser,
    response: measured_entity_class_metrics_serialiser_1.MeasuredEntityClassMetricsSerialiser
};
exports.PUT_TOPIC_METRIC_COLLECTOR = {
    id: 146,
    request: topic_metric_collector_serialiser_1.TopicMetricCollectorSerialiser,
    response: NullSerialiser
};
exports.REMOVE_TOPIC_METRIC_COLLECTOR = {
    id: 147,
    request: string_serialiser_1.StringSerialiser,
    response: NullSerialiser
};
exports.FETCH_TOPIC_METRIC_COLLECTORS = {
    id: 148,
    request: NullSerialiser,
    response: topic_metric_collector_list_serialiser_1.TopicMetricCollectorListSerialiser
};
exports.FETCH_JMX_VALUES = {
    id: 149,
    request: jmx_object_names_serialiser_1.JMXObjectNamesSerialiser,
    response: mbean_values_serialiser_1.MBeanValuesSerialiser
};
exports.CREATE_REMOTE_SERVER = {
    id: 157,
    request: remote_server_definition_serialiser_1.RemoteServerDefinitionSerialiser,
    response: create_remote_server_result_serialiser_1.CreateRemoteServerResultSerialiser
};
exports.REMOVE_REMOTE_SERVER = {
    id: 158,
    request: string_serialiser_1.StringSerialiser,
    response: NullSerialiser
};
exports.UPDATE_LICENSE = {
    id: 159,
    request: bytes_serialiser_1.BytesSerialiser,
    response: replace_licence_response_serialiser_1.ReplaceLicenceResponseSerialiser
};
exports.CLOSE_CLIENT_FILTER = {
    id: 162,
    request: string_serialiser_1.StringSerialiser,
    response: count_or_parser_errors_serialiser_1.CountOrParserErrorsSerialiser
};
exports.SET_CLIENT_CONFLATION_FILTER = {
    id: 163,
    request: set_client_conflation_filter_request_serialiser_1.SetClientConflationFilterRequestSerialiser,
    response: count_or_parser_errors_serialiser_1.CountOrParserErrorsSerialiser
};
exports.LIST_REMOTE_SERVERS = {
    id: 164,
    request: NullSerialiser,
    response: list_remote_servers_result_serialiser_1.ListRemoteServersResultSerialiser
};
exports.CHECK_REMOTE_SERVER = {
    id: 165,
    request: string_serialiser_1.StringSerialiser,
    response: check_remote_server_result_serialiser_1.CheckRemoteServerResultSerialiser
};
