// diffusion/src/client/client-summary.js
"use strict";
/**
 * @module diffusion.clients
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Enum containing possible client types.
 */
var ClientType;
(function (ClientType) {
    /**
     * JavaScript client
     */
    ClientType[ClientType["JAVASCRIPT_BROWSER"] = 0] = "JAVASCRIPT_BROWSER";
    /**
     * Android client
     */
    ClientType[ClientType["ANDROID"] = 3] = "ANDROID";
    /**
     * iOS client
     */
    ClientType[ClientType["IOS"] = 4] = "IOS";
    /**
     * Java client
     */
    ClientType[ClientType["JAVA"] = 8] = "JAVA";
    /**
     * .NET client
     */
    ClientType[ClientType["DOTNET"] = 9] = "DOTNET";
    /**
     * C
     */
    ClientType[ClientType["C"] = 10] = "C";
    /**
     * Client type is not known to the local session, possibly because the remote client is using a different version
     * of Diffusion
     */
    ClientType[ClientType["OTHER"] = 12] = "OTHER";
})(ClientType = exports.ClientType || (exports.ClientType = {}));
/**
 * Enum containing possible Transport types.
 */
var TransportType;
(function (TransportType) {
    /**
     * WebSocket protocol
     */
    TransportType[TransportType["WEBSOCKET"] = 0] = "WEBSOCKET";
    /**
     * HTTP long polling transport
     */
    TransportType[TransportType["HTTP_LONG_POLL"] = 1] = "HTTP_LONG_POLL";
    /**
     * HTTP long poll, via HTML iFrame elements
     */
    TransportType[TransportType["IFRAME_LONG_POLL"] = 2] = "IFRAME_LONG_POLL";
    /**
     * HTTP 1.1 chunked transfer via HTML iFrame elements
     */
    TransportType[TransportType["IFRAME_STREAMING"] = 3] = "IFRAME_STREAMING";
    /**
     * Diffusion protocol based on TCP sockets
     */
    TransportType[TransportType["DPT"] = 4] = "DPT";
    /**
     * Diffusion protocol based on HTTP 1.1 chunked transfer
     */
    TransportType[TransportType["HTTPC"] = 5] = "HTTPC";
    /**
     * Diffusion protocol based on HTTP 1.1 chunked transfer
     */
    TransportType[TransportType["HTTPC_DUPLEX"] = 6] = "HTTPC_DUPLEX";
    /**
     * Transport type is not known to the local session, possible because the remote client is using a different version
     * of Diffusion
     */
    TransportType[TransportType["OTHER"] = 7] = "OTHER";
})(TransportType = exports.TransportType || (exports.TransportType = {}));
