// services/wills/will-registration-result.js
"use strict";
/**
 * @module Services.Wills
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The result of a will registration
 */
var WillRegistrationResultEnum;
(function (WillRegistrationResultEnum) {
    WillRegistrationResultEnum[WillRegistrationResultEnum["SUCCESS"] = 0] = "SUCCESS";
    WillRegistrationResultEnum[WillRegistrationResultEnum["GROUP_CONFLICT"] = 1] = "GROUP_CONFLICT";
    WillRegistrationResultEnum[WillRegistrationResultEnum["TOPIC_TREE_CONFLICT"] = 2] = "TOPIC_TREE_CONFLICT";
})(WillRegistrationResultEnum = exports.WillRegistrationResultEnum || (exports.WillRegistrationResultEnum = {}));
/**
 * Server result from a TOPIC_SCOPED_WILL_REGISTRATION request.
 */
var WillRegistrationResult = /** @class */ (function () {
    /**
     * Create a new WillRegistrationResult instance
     *
     */
    function WillRegistrationResult(result) {
        this.result = result;
    }
    return WillRegistrationResult;
}());
exports.WillRegistrationResult = WillRegistrationResult;
